Tweak E0599 and elaborate_predicates

estebank opened the pull request:
CC https://github.com/rust-lang/rust/issues/86377.

estebank committed:
Elaborate unmet obligations in E0599 for more context

estebank committed:
Keep obligation chain when elaborating obligations

estebank committed:
Add tests

estebank commented:
r? @compiler-errors

compiler-errors commented on compiler/rustc_hir_typeck/src/method/probe.rs:
```diff
@@ -1587,11 +1587,29 @@ impl<'a, 'tcx> ProbeContext<'a, 'tcx> {
+                                let parent = if o == parent_o {
+                                    None
+                                } else {
+                                    if o.predicate.to_opt_poly_trait_pred().map(|p| p.def_id())
+                                        == self.tcx.lang_items().sized_trait()
```
```suggestion
                                    if let Some(trait_pred) = o.predicate.to_opt_poly_trait_pred()
                                        && self.tcx.lang_items().sized_trait() == Some(trait_pred.def_id())
```

This will never happen in practice, but it's better to deal with the `None == None` case when lang-items are not defined.

estebank replied:
What is the `None == None` case you envision? Because `trait_pred` might have been `None` (so we don't get into this logic) and `sized_trait()` might be `None` which might cause a build where the sized trait lang item isn't present to be too verbose (with extra labels), no other behavioral change.

compiler-errors commented on tests/ui/missing-trait-bounds/issue-35677.fixed:
```diff
@@ -3,7 +3,7 @@
 use std::hash::Hash;
 
-fn is_subset<T>(this: &HashSet<T>, other: &HashSet<T>) -> bool where T: Eq, T: Hash {
+fn is_subset<T>(this: &HashSet<T>, other: &HashSet<T>) -> bool where T: Eq, T: Hash, T: PartialEq {
```
Hm, I don't consider this an improvement -- `T: Eq` implies `T: PartialEq` by elaboration.

estebank replied:
I agree, so I tested hiding the suggestions when a parent obligation was present, but that broke some existing tests with applicable suggestions. I'll see if we can inspect the obligation cause code instead to avoid these.

estebank replied:
Fixed

compiler-errors commented:
Ideally we'd distinguish elaborated predicates and "root" predicates so that we only pass the latter for suggestions, but I guess whatever :thinking: 

r=me

estebank committed:
Do not incorrectly suggest restricting implied bounds

When we have already suggested bounds that imply the about to be
suggested bound, skip them.

estebank commented:
@bors r=compiler-errors

bors commented:
:pushpin: Commit 22a0e4fa6e08acf2de50ed87cfb909092ab0459d has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 22a0e4fa6e08acf2de50ed87cfb909092ab0459d -->
<!-- homu: {"type":"Approved","sha":"22a0e4fa6e08acf2de50ed87cfb909092ab0459d","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
